### PR TITLE
[#10146] Result Page: Add indicates missing response features

### LINKED ISSUE #10146: Result Page: Add indicates missing response features
image

The task has been divided into two phases:

- **(1)** Due the legacy complicated implementation of fetching session result bundle, the backend logic need to refactored and clean to suit V7. In V7, the backend should not care about data presentation and therefore we can write a much clean logic.

- **(2)** Add indicates missing response in the backend in the frontend (WIP).

### PR DESCRIPTION
**Pull request by xpdavid**
Fixes #10146 

image

image

image

The PR also fixes several bugs of missing responses feature in V6 where the giver recipient pairs are not complete.

For example, missing responses for anonymous student will be shown:
image

**Commit Organization**
- The first two commits are just refactor work and are explained in the commit message
- The third commit enables the generation of missing responses in the backend. Here is the procedure (computation heavy):
    1. We generate all giver-recipient pairs for questions
    2. Remove pairs that have real corresponding responses
    3. Build missing responses and ship them in the `allResponses` field of `SessionResultData`. Missing responses are indicated by a new field in `ResponseOutput` called `isMissingResponse`.
- The last commit displays the missing responses in the frontend. 

There is one potential performance concern where the missing responses are generated in the backend without using boolean variable (reflected as HTTP params) to control the generation. 

There are several reasons there:
- The generation of missing responses are purely computational works and the performance is only related to how big the course is.
- We don't need to resend the same result API request just to retrieve the missing responses and thus result in more clean code in the frontend
- Don't do Premature Optimization. The check box is checked by default when the page is loaded so the generation will always be run anyway.

### CODE REVIEW

File: src/test/java/teammates/test/cases/datatransfer/SessionResultsBundleTest.java [hunk 1]
```diff
@@ -23,7 +23,7 @@
 public class SessionResultsBundleTest extends BaseTestCase {
 
     @Test
-    public void testGetResponseCommentsMap() {
+    public void testGetQuestionResponseMapMap() {
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/datatransfer/SessionResultsBundleTest.java:26]: `MapMap` -> `Map`

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java [hunk 2]
```diff
@@ -2019,7 +2037,7 @@ public void testGetSessionResultsForUser_instructorAllQuestions_shouldGenerateCo
                 getResponseId("qn5.resp1", responseBundle) + "={false,true}",
                 getResponseId("qn6.resp1", responseBundle) + "={true,true}");
         AssertHelper.assertContains(expectedStrings, mapString);
-        assertEquals(10, bundle.getResponseVisibilityTable().size());
+        assertEquals(58, bundle.getResponseVisibilityTable().size());
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java:2040]: 58 = 10 (total response) + 48 (total missing). It's not easy to justify where 10 and 48 come from, but 58 should not be left as a magic number.

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java [hunk 3]
```diff
@@ -2061,7 +2086,7 @@ public void testGetSessionResultsForUser_instructorAllQuestionsSpecificSection_s
                 getResponseId("qn2.resp3", responseBundle) + "={false,false}",
                 getResponseId("qn2.resp1", responseBundle) + "={false,false}");
         AssertHelper.assertContains(expectedStrings, mapString);
-        assertEquals(7, bundle.getResponseVisibilityTable().size());
+        assertEquals(43, bundle.getResponseVisibilityTable().size());
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java:2089]: Similarly here for 43 = 7 + 36.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 4]
```diff
@@ -1623,6 +1632,92 @@ private void addCommentVisibilityToTable(
...
+                    }
+
+                    FeedbackResponseAttributes missingResponse =
+                            FeedbackResponseAttributes.builder(questionId, giverIdentifier, recipientIdentifier)
+                                    .withCourseId(feedbackSession.getCourseId())
+                                    .withFeedbackSessionName(feedbackSession.getFeedbackSessionName())
+                                    .withGiverSection(giverInfo.getSectionName())
+                                    .withRecipientSection(recipientInfo.getSectionName())
+                                    .withResponseDetails(new FeedbackTextResponseDetails(
+                                                    Const.INSTRUCTOR_FEEDBACK_RESULTS_MISSING_RESPONSE))
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java:1699]: Is it possible/not too troublesome to generate the response details object based on the question type?

Comment (xpdavid) [reply src/main/java/teammates/logic/core/FeedbackSessionsLogic.java:1699]: Actually it is hard. There is no "default value factory" in the backend (but there is one in the frontend)

Comment (wkurniawan07) [reply src/main/java/teammates/logic/core/FeedbackSessionsLogic.java:1699]: Then it's ok to let it be like this

**Pull request merged**